### PR TITLE
Probability Samplers based on W3C Trace Context Level 2

### LINKED ISSUE #1413: Review approach & specify algorithm for TraceIdRatioBasedSampler (ProbabilitySampler)
See also discussion on https://github.com/open-telemetry/opentelemetry-specification/pull/1412#issuecomment-776101910

The sampling alorithm for TraceIdRatioBasedSampler is unspecified. As a result, trace IDs that are sampled by some implementations might get non-sampled or re-sampled by SDKs in other languages, even though they have the same or a a higher probability than the parent.

TODO list for this issue:

1. Is this a problem at all? Or is the ParentBased approach enough (in combination with any, not necessarily trace-id-based probability-based sampling of root spans)?
2. ~~If it is a problem, since trace IDs can come from untrusted, non-random sources, do we open up a DDoS/Security/performance issue when using trace IDs as sole, deterministic input for our sampling algorithm? Do we need to put a warning there?~~ Let's assume it is no problem for this issue, this should be handled in #1188.
3. If applicable after checking the above we determine that we need a consistent algorithm, actually do specify it (maybe based on #331).

### LINKED ISSUE #3602: Decision on new encoding for sampling "selectivity"
## Summary

The Sampling SIG has been working on a proposal to follow the W3C tracecontext group, which has added a flag to convey definite information about randomness in the TraceID.

In particular, we aim to address the [TODO](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#traceidratiobased) about the TraceIDRatioBased Sampler:

> TODO: Add details about how the TraceIdRatioBased is implemented as a function of the TraceID. #1413

We are looking for community input on a choice which will impact implementation complexity for OpenTelemetry Samplers as well as human-interpretability of the raw data.

Note this proposal was co-authored by @kentquirk @oertl @PeterF778 and @jmacd. 

**ACTION ITEM**: Please review and vote for your preferred encoding strategy in the comments below. OpenTelemetry Tracing SDK authors as well as OpenTelemetry Collector trace processors will be asked to implement the encoding and decoding strategies here in order to communicate about sampling selectivity, and we need your input!

## Background

We propose to add information to the W3C Trace Context specification to allow consistent sampling decisions to be made across the entire lifetime of a trace.

The expectation is that trace IDs should contain at least 56 bits of randomness in a known portion of the ID. This value is known as r, and there is a bit in the trace header that indicates its presence.

In probabilistic sampling, the sampling decision is a binary choice to keep (store) or drop (discard) a trace. Because traces are composed of multiple spans, we want to be sure that the same decision is made for all elements in the trace. Therefore, we don’t make a truly random decision at each stage. We instead wish to use the randomness embedded in the trace ID so that all stages can make consistent decisions.

In order to make consistent decisions, we need to propagate not only the randomness (the r value), but also the sampling selectivity used. In other words, in a trace that travels between services A, B, and C, any decision made by B should use the same information as a decision made by A, and B could potentially modify the selectivity so that C could also make an appropriate decision.

As noted, the r value expresses a 56-bit random value that can be used as the source of randomness for a probabilistic sampling decision. The intent of this proposal is to express the sampling selectivity that was used to make the decision, and to do it in the trace state.

Sampling selectivity can be described in several different ways:
- As a probability – a value between 0 (100% chance of being dropped) and 1 (100% chance of being kept). 
- As a sampling rate – an indication of the ratio between the total number of items and the number of items kept. It is typically a positive integer. It is the inverse of the probability – a sampling rate of 8 is equivalent to a sampling probability of 0.125.
- As a threshold – if r (56 bits of randomness) is treated as a 56-bit integer between 0 and (2^56)-1, then the threshold is the limit below which the sample will be kept. For example, the sampling rate of 8 is equivalent to a threshold value of 2^53, or 0x20000000000000.

## Minimum requirements

Given the sampling information on the trace state it MUST be specified for any possible representations on any platform, how this translates to the applied sampling threshold (the value that was used to compare against the random bits). Only this allows to reproduce the sampling decision together with the 56 random bits and gives 100% consistency.

Based on that, it can be derived which of the 2^56+1 sampling thresholds, that are meaningful when having 56 random bits, can be expressed by the sampling information on the trace state. The proposals should therefore be clear about which thresholds are actually supported. The set of supported thresholds also defines the set of possible sampling probabilities. The sampling probability is just the threshold multiplied by 2^(-56).

When picking one of the supported thresholds, there should be a lossless way to map it to the sampling information that is written to the trace state. Lossless in the sense, that the reverse mapping as described in 1. yields again exactly the chosen threshold. The mapping from thresholds to the sampling information is important for adaptive sampling, where the threshold is automatically chosen.

## Objective

We would like to express this sampling probability/rate/threshold in a reasonably compact way in the trace state. We would like that expression to be easy and efficient to generate and parse in any programming language. Another requirement is that the used notation should be able to describe cases of non-probabilistic sampling (corresponding to the zero adjusted count or the old p=63 case). We have been referring to this value as t.

The sampling SIG has been discussing this issue for some time, and we have examined several proposals. Each proposal has its strengths and weaknesses and we are looking for input from the larger community.

Note that this document includes just a summary of the proposals, below; all of them have been specified in sufficient detail to resolve most implementation issues. We are hoping for input from the community to help make the big decision about the implementation direction.

## Request for input

The major difference in these proposals that we wish to seek input on is whether it is more important to optimize for threshold calculation (option A) at the expense of human readability, or whether to choose one of the other options which are readable and accessible, but make threshold calculations harder to work with.

## List of options

When we refer to Tmax, we mean 2^56 (0x100000000000000 or 72057594037927936)

### Option A: Hex Threshold

<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-ce002662-7fff-e1fb-be13-2963bff5c36b"><div dir="ltr" style="margin-left:0pt;" align="left">

Title | A: Hex Threshold
-- | --
Description | The t value is expressed as a threshold value in the range [0, 2^56-1] using 14 hexadecimal digits; an absent t-value represents a threshold of 2^56 corresponding to 100% sampling probability
Examples | Keep all: t is omitted<br>Keep 1 in 10: t=19999999999999<br>Keep 1 in 8: t=20000000000000<br>Keep half: t=80000000000000<br>Keep 2/3: t=aaaaaaaaaaaaaa<br>Keep 1 in 1000: t=004189374bc6a7
Mapping to the sampling threshold | threshold = parseHexInt(t)<br>If t is absent the threshold is 2^56
Supported thresholds | All 2^56 + 1 thresholds that are meaningful when using 56 random bits. Possible thresholds are {0, 1, 2, 3, …., 2^56-1, 2^56}.
Mapping a supported sampling threshold to t | t = encodeHexInt(threshold)<br>If t is 2^56, corresponding to 100% sampling probability, the t-value is not set
Advantages | Both t and r can be parsed as hex strings using the same parsing function and compared directly with no further processing. Simplest approach satisfying the minimum requirements above. t-value can be compared by humans to trace-ID to understand the sampling decision. No floating-point operations or complex parsing needed.
Disadvantages | t-value cannot be directly read by humans as sampling probability.

### Option A<sup>1</sup>: Hex Threshold with omission of trailing zeros

<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-9bba75da-7fff-441a-be3c-3efb8c15385f"><div dir="ltr" style="margin-left:0pt;" align="left">

Title | A1: Hex Threshold (omission of trailing zeros allowed)
-- | --
Description | The t value is expressed as a threshold value in the range [0, 2^56-1] using 14 hexadecimal digits; an absent t-value represents a threshold of 2^56 corresponding to 100% sampling probability, trailing 0s may be omitted for brevity.
Examples | Keep all: t is omitted<br>Keep 1 in 10: t=19999999999999<br>Keep 1 in 8: t=2<br>Keep half: t=8<br>Keep 2/3: t=aaaaaaaaaaaaaa<br>Keep 1 in 1000: t= 004189374bc6a7
Mapping to the sampling threshold | threshold = parseHexInt(t)<br>If t is absent the threshold is 2^56t is padded with zeros if it has less than 14 hex digits
Supported thresholds | All 2^56 + 1 thresholds that are meaningful when using 56 random bits
Mapping a supported sampling threshold to t | t = encodeHexInt(threshold)<br>If t is 2^56, corresponding to 100% sampling probability, the t-value is not settrailing zeros may be omitted
Advantages | Both t and r can be parsed as hex strings and compared directly with no further processing. Simplest approach satisfying the minimum requirements above. t-value can be compared by humans to trace-ID to understand the sampling decision. No floating-point operations or complex parsing needed.Compact representation for certain probabilities, especially those that are powers of two
Disadvantages | t-value cannot be directly read by humans as sampling probability. compact power-of-2 representations are almost misleading. Other common sample rates are not expressed in compact or obvious ways.

</div></b>

## Option B: Integer Sampling Rate

<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-2afb25f0-7fff-8eab-9fbf-d252dbfe925f"><div dir="ltr" style="margin-left:0pt;" align="left">

Title | B: Integer Sampling Rate
-- | --
Description | The t value is expressed as a positive integer representing the ratio between the number of items kept and the total number of items.
Examples | Keep all: t=1<br>Keep 1 in 8: t=8<br>Keep 1 in 10: t=10<br>Keep half: t=2<br>Keep 2/3: not expressible in this format<br>Keep 1 in 1000: t=1000<br>Keep none: not expressible in this format
Mapping to the sampling threshold | threshold = round(2^56 / parseDecimalInt(t))
Supported thresholds | 2^56, 2^56/2, 2^56/3, 2^56/4, …. corresponding to sampling probabilities 1, ½, ⅓, ¼, …
Mapping a supported sampling threshold to t | t = encodeDecimalInt(round(2^56 / threshold)), this reverse mapping can be lossless for integers up to 268M
Advantages | Easy format for most common rates. Adjusted counts (extrapolation factors) are guaranteed to be integers.
Disadvantages | There are many values it can’t express; particularly the desire to keep more than half but less than all of the data. Mappings require floating-point divisions.

</div></b>

## Option C: Sampling probability

<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-3095d485-7fff-d028-3504-1b3210a6d19d"><div dir="ltr" style="margin-left:0pt;" align="left">

Title | C: Sampling Probability
-- | --
Description | The t value is expressed as a positive decimal floating point value between 0 and 1, representing the probability of keeping a given event.
Examples | Keep all: t=1<br>Keep 1 in 8: t=.125<br>Keep 1 in 10: t=.1<br>Keep half: t=.5<br>Keep 2/3 ieee precision: t=.6666666666667<br>Keep 2/3 precision 4: t=.6667<br>Keep 2/3 precision 2: t=.67<br>Keep 1 in 1000: t=.001
Mapping to the sampling threshold | threshold = 2^56 * parseDecimalFloat(t)<br>Note rounding is performed in parseDecimalFloat.
Supported thresholds | If t is a double-precision floating-point number, at least all thresholds with 3 least significant bits being equal to zero.
Mapping a supported sampling threshold to t | t = encodeDecimalFloat(threshold * 2^(-56)), this reverse mapping can be lossless for all supported thresholds when using enough decimal digits
Advantages | Easy format for humans to read.
Disadvantages | Requires floating point math. Doesn’t express powers of 2 compactly.

</div></b>

### Option C<sup>1</sup>: Sampling probability with hex floating point

<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-254aa3a6-7fff-9797-6f6f-55a4429a4831"><div dir="ltr" style="margin-left:0pt;" align="left">

Title | C<sup>1</sup>: Sampling Probability with hex floating point
-- | --
Description | The t value is expressed as a positive decimal or hexadecimal floating point value between 0 and 1, representing the probability of keeping a given event. The encoding MAY use C99 and IEEE-754-2008 specified hex floating point as an exact representation.Note about precision: IEEE-754 double-wide floating point numbers carry 52 bits of significand, 4 bits fewer than trace randomness. Below, “ieee precision” refers to 53 bits of precision.
Examples | Keep all: as in C<br>Keep 1 in 8: as in C or t=0x1p-3<br>Keep 1 in 10: as in C or 0x1.5p-3<br>Keep half: as in C or t=0x1p-1<br>Keep 2/3 ieee precision: as in C or 0x1.5555555555555p-1 <br>Keep 2/3 precision 4: as in C or 0x1.5555p-1 <br>Keep 2/3 precision 2: as in C or 0x1.55p-1<br>Keep 1 in 1000 ieee precision: as in C or 0x1.0624dd2f1a9fcp-10<br>Keep 1 in 1000 precision 4: as in C or 0x1.0625p-10<br>Keep 1 in 1000 precision 2: as in C or 0x1.06p-10
Mapping to the sampling threshold | threshold = 2^56 * parseHexFloat(t)<br>Note there is no rounding performed.
Supported thresholds | If t is a double-precision floating-point number, at least all thresholds with 3 least significant bits being equal to zero.
Mapping a supported sampling threshold to t | t = encodeHexFloatFromThreshold(threshold), this method is exact and lossless using built-in libraries up to 52 bits of precision, and exact and lossless at 56 bits of precision using custom code.
Advantages | Decimal format for humans to read (as in C), but hex format permits exact and lossless encoding of arbitrary thresholds up to 56 bits
Disadvantages | Users require custom code, or floating point math and a library that supports hex representation of it. Since this was added in ISO-C99 (1999) and IEEE-754 (2008), it is relatively widespread.

## Option C<sup>2</sup>: Sampling Probability with unnormalized hex floating point

<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-1fe08bf2-7fff-df8d-9b07-fcc95734fc1b"><div dir="ltr" style="margin-left:0pt;" align="left">

Title | C<sup>2</sup>: Sampling Probability with unnormalized hex floating point
-- | --
Description | In addition to C and C1, encoders are encouraged to use unnormalized hex floating point when they synthesize arbitrary probabilities, because it is lossless and can be easily read as a sampling threshold by humans.
Examples | Keep all: as in C1<br>Keep 1 in 8: as in C1 or t=0x2p-04 (threshold = 0x20000000000000)<br>Keep 1 in 10: as in C1 or t=0x2ap-8 (threshold = 0x2a000000000000)<br>Keep half: as in C1 or t=0x8p-04 (threshold 0x80000000000000)<br>Keep 2/3 full precision: as in C1 or 0xaaaaaaaaaaaaaap-56 <br>Keep 2/3 ieee precision: as in C1 or 0xaaaaaaaaaaaaap-52<br>Keep 2/3 precision 4: as in C1 or 0xaaabp-16<br>Keep 2/3 precision 2: as in C1 or 0xabp-8<br>Keep 1 in 1000 ieee precision: as in C1 or 0x4189374bc6a7p-56<br>Keep 1 in 1000 precision 4: as in C1 or 0x4189p-16<br>Keep 1 in 1000 precision 2: as in C1 or 0x42p-8
Mapping to the sampling threshold | threshold = 2^56 * parseHexFloat(t)<br>Note there is no rounding performed.
Supported thresholds | If t is a double-precision floating-point number, at least all thresholds with 3 least significant bits being equal to zero.
Mapping a supported sampling threshold to t | t = encodeHexFloatFromThreshold(threshold), this method is exact and lossless using built-in libraries up to 52 bits of precision, and exact and lossless at 56 bits of precision using custom code.
Advantages | Decimal format for humans to read (as in C), but hex format permits exact and lossless encoding of arbitrary thresholds up to 56 bits
Disadvantages | Users require custom code, or floating point math and a library that supports hex representation of it. Since this was added in ISO-C99 (1999) and IEEE-754 (2008), it is relatively widespread.

## Option D: Combination of C and D

<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-cae57b17-7fff-755d-4b54-f352b9818634"><div dir="ltr" style="margin-left:0pt;" align="left">

Title | D: Combine B and C<sup>21</sup>
-- | --
Description | The t value can be either a value <1, in which case it’s interpreted as in C (or C2 preferably or C1) above. Or it can be >=1, in which case it’s interpreted as in B. Implementations are recommended to limit precision to keep the encoding of sampling probabilities compact.
Examples | Keep all: t=1<br>Keep 1 in 8: t=8 or t=.125<br>Keep 1 in 10: t=.1 or t=10<br>Keep half: t=.5 or t=2<br>Keep 2/3: t=.6667<br>Keep 1 in 1000: t=.001 or t=1000<br>Keep arbitrary hex-digit threshold HHHH with custom code: t=0xHHHHp-(len(HHHH)*4)<br>Keep arbitrary hex-digit threshold HHHH with standard library:t=0x1.JJJJJp-DDwhere JJJJJ and DD correspond with the normalized hex floating point value value corresponding with HHHH. Note that JJJJJ is one digit longer than HHHH, due to shifting hex digits by one bit.
Mapping to the sampling threshold | See B and C
Supported thresholds | The union of B and C. Hence, if t is a double-precision floating-point number, at least all thresholds with 3 least significant bits being equal to zero.
Mapping a supported sampling threshold to t | Dependent on whether the threshold is supported by B or C, one has to choose the reverse mapping of B or C, respectively.
Advantages | The most convenient representation can be used. This allows constant user-input probabilities to be represented exactly in their original human-readable form, while it allows machine-generated probabilities to be rounded and losslessly encoded with variable precision.
Disadvantages | Requires floating point math or custom code. Parsing is slightly more complex.

</div></b>

## Option E: Ratio

<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-625959e9-7fff-19ee-64a8-4d0db566fc85"><div dir="ltr" style="margin-left:0pt;" align="left">

Title | E: Ratio
-- | --
Description | The t value is expressed as a ratio of two integers n and d, separated by a slash. If n is 1 it may be omitted.
Examples | Keep all: t=1<br>Keep 1 in 8: t=8 or t=1/8<br>Keep 1 in 10: t=10 or t=1/10<br>Keep half: t=50/100 or t=1/2 or t=2<br>Keep 2/3: t=2/3 or t=6667/10000<br>Keep 1 in 1000: t=1/1000 or t=1000
Mapping to the sampling threshold | threshold = parseInt(t_numerator) * 2^56 / parseInt(t_denominator)(this formula may overflow when using 64-bit integers)
Supported thresholds | The set of supported thresholds cannot be easily described and depends on the value range of the numerator and the denominator.
Mapping a supported sampling threshold to t | Not unique and difficult as it relates to finding the closest ratio to the value given by threshold * 2^(-56)
Advantages | Does not require floating point or complex parsing but can express a full range of sample rates.
Disadvantages | Converting from a probability to a ratio may lose precision.

</div></b>

## Option F: Powers-of-two

<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-00716d1f-7fff-aed0-5087-8983ae51e5c9"><div dir="ltr" style="margin-left:0pt;" align="left">

Title | F:Powers of 2
-- | --
Description | The t value is expressed as the exponent of the sampling probability given by 0.5^t, see current proposal https://opentelemetry.io/docs/specs/otel/trace/tracestate-probability-sampling/#consistent-probability-sampling
Examples | Keep all: t=0<br>Keep 1 in 8: t=3<br>Keep half: t=1
Mapping to the sampling threshold | threshold = 2^(56 - parseInt(t))
Supported thresholds | 2^56, 2^55, 2^54,…., 1 corresponding to sampling probabilities 1,½, ¼,...,2^(-56)..
Mapping a supported sampling threshold to t | t = encodeInt(numberOfLeadingZeros(threshold)-7), provided that the threshold is a 64-bit integer numberOfLeadingZeros requires a single CPU instruction
Advantages | Simple and compact. Adjusted counts (extrapolation factors) are guaranteed to be integers. Fast mapping to thresholds and vice versa.
Disadvantages | Only supports power of 2 sampling probabilities.

</div></b>

### PR DESCRIPTION
**Pull request by jmacd**
Fixes #1413.
Fixes #3602.

## Changes

Updates Trace SDK and Propagator specifications with complete support for consistent probability sampling as described in [OTEP 235](https://github.com/open-telemetry/oteps/blob/main/text/trace/0235-sampling-threshold-in-trace-state.md).

* [ ] Related issues #3307, #2253, #2179, #2113, #1947,#1844,
* [ ] OTEP: https://github.com/open-telemetry/oteps/pull/235
* [ ] Links to the prototypes https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/29720
* [ ] [`CHANGELOG.md`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/CHANGELOG.md) TODO
* [ ] [`spec-compliance-matrix.md`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/spec-compliance-matrix.md) TODO

### CODE REVIEW

**Comment by jmacd** [comment]
@kentquirk @oertl @PeterF778 @kalyanaj Please take a look.
I believe a bit more prototyping will be required, so I leave this open as a draft PR.
I am looking to prototype an OTel-Go SDK change to the TraceIdRatioBased sampler, and I will be continuing to revive https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/24811  in the coming week. Cheers!

**Comment by jmacd** [comment]
@kentquirk @oertl @PeterF778 @kalyanaj 
TL;DR I found about half of the text in the old, experimental specification to be overkill and removed a lot of it. If you see something you'll greatly miss, we can bring it back. @kentquirk if you want to throw out the old, experimental specification and start from scratch, I'm open to that. I don't think we need to extend the specification with all the pseudocode from the OTEP, readers can refer back to that document for details.

File: specification/context/api-propagators.md [hunk 1]
```diff
@@ -355,6 +355,22 @@ Additional `Propagator`s implementing vendor-specific protocols such as AWS
...
+
+When injecting and extracting trace context to ro from a carrier, the
+following fields are propagated.
+
+- TraceID (16 bytes)
+- SpanID (8 bytes)
+- TraceFlags (8 bits)
+- TraceState (string)
+
+Propagators MUST NOT assume that bits 2-7 (6 most significant bits) will be zero.
```
Comment (PeterF778) [inline specification/context/api-propagators.md:372]: For clarity, let's say "bits 2-7 of TraceFlags".

File: specification/trace/sdk.md [hunk 2]
```diff
@@ -312,21 +333,46 @@ The default sampler is `ParentBased(root=AlwaysOn)`.
...
+which is based on the [W3C Trace Context Level 2 Random TraceID
+flag](https://www.w3.org/TR/trace-context-2/#random-trace-id-flag).
+When a TraceID has the `Random` bit set, samplers are able to use 56
+specific bits of consistent randomness for sampling decisions.
+
+The implementation has the following steps:
+
+* Ratio values are restricted to the range `2**-56` through 1
+* A rejection threshold is calculated, expressing as an integer how
+  many out of 2**56 trace IDs should be sampled.
```
Comment (PeterF778) [inline specification/trace/sdk.md:353]: "sampled" => "dropped"

File: specification/trace/tracestate-probability-sampling.md [hunk 3]
```diff
@@ -295,432 +286,355 @@ applies to the two samplers specified here as well as consumers of
...
 
-P-value and r-value are set in the OpenTelemetry `tracestate`, under
-the vendor tag `ot`, using the identifiers `p` and `r`.  P-value is an
-unsigned integer valid in the inclusive range `[0, 63]` (i.e., there
-are 64 valid values).  R-value is an unsigned integer valid in the
-inclusive range `[0, 62]` (i.e., there are 63 valid values).  P-value
-and r-value are independent settings, each can be meaningfully set
-without the other present.
+The [Sampler API](sdk.md#sampler) is responsible for setting the
+`sampled` and `random` flags and the `tracestate`.
```
Comment (PeterF778) [inline specification/trace/tracestate-probability-sampling.md:306]: This is problematic. A sampler should not change the `random` flag at all. This flag must be set when the TraceID is created which is ahead of making the sampling decision for the root span.

File: specification/trace/tracestate-probability-sampling.md [hunk 4]
```diff
@@ -295,432 +286,355 @@ applies to the two samplers specified here as well as consumers of
...
-
-##### Requirement: R-value is generated with the correct probabilities
-
-Samplers MUST generate r-values using a randomized scheme that
-produces each value with the probabilities equivalent to those
-produced by counting the number of leading 0s in a string of 62 random
-bits.
+Head Samplers (i.e., trace SDKs) SHOULD by default generate the 56
+random bits specified for TraceIDs by the W3C Trace Context Level 2,
+and set the Random trace flag, when generating new contexts.
```
Comment (PeterF778) [inline specification/trace/tracestate-probability-sampling.md:367]: According to the SDK specification, the TraceID is already generated even when sampling ROOT spans. This needs more discussions.

File: specification/context/api-propagators.md [hunk 5]
```diff
@@ -355,6 +355,22 @@ Additional `Propagator`s implementing vendor-specific protocols such as AWS
 X-Ray trace header protocol MUST NOT be maintained or distributed as part of
 the Core OpenTelemetry repositories.
 
+### W3C Trace Context Requirements
+
+A W3C Trace Context propagator is expected to implement the
+`traceparent` and `tracestate` contexts fields specified in [W3C Trace
```
Comment (kentquirk) [inline specification/context/api-propagators.md:361]:
```suggestion
`traceparent` and `tracestate` context fields specified in [W3C Trace
```

File: specification/context/api-propagators.md (hunk 1 above)
Comment (kentquirk) [inline specification/context/api-propagators.md:372]:
```suggestion
Propagators MUST NOT assume that bits 2-7 of TraceFlags (the 6 most significant bits) will be zero.
```

File: specification/trace/api.md [hunk 6]
```diff
@@ -221,7 +221,10 @@ byte.
 
 `TraceFlags` contain details about the trace. Unlike TraceState values,
 TraceFlags are present in all traces. The current version of the specification
-only supports a single flag called [sampled](https://www.w3.org/TR/trace-context/#sampled-flag).
+supports two flags:
+
+- [Sampled](https://www.w3.org/TR/trace-context/#sampled-flag)
+- [Random](https://www.w3.org/TR/trace-context-2/#random-trace-id-flag)
```
Comment (kentquirk) [inline specification/trace/api.md:227]:
```suggestion
- [Sampled (value 0x1)](https://www.w3.org/TR/trace-context/#sampled-flag)
- [Random (value 0x2)](https://www.w3.org/TR/trace-context-2/#random-trace-id-flag)
```
I think it would help to know which bits we're talking about.

File: specification/trace/sdk.md (hunk 2 above)
Comment (kentquirk) [inline specification/trace/sdk.md:353]:
```suggestion
* A rejection threshold is calculated, expressing as an integer the number
  out of 2**56 trace IDs that should be dropped.
```

File: specification/trace/sdk.md [hunk 7]
```diff
@@ -460,6 +507,30 @@ Additional `IdGenerator` implementing vendor-specific protocols such as AWS
...
+### Randomness requirements
+
+The SDK SHOULD implement the TraceID randomness requirements specified
+in the W3C [Trace Context Level
+2](https://www.w3.org/TR/trace-context-2/#randomness-of-trace-id)
+Candidate Recommendation.
+
+This states that the SDK should fill least significant 7 bytes (i.e., 56
+bits) of the TraceID are genuinely random or pseudo-random., so they
+can be used for probability sampling.
```
Comment (kentquirk) [inline specification/trace/sdk.md:519]:
```suggestion
This states that the SDK should fill the least significant 7 bytes (i.e., 56
bits) of the TraceID with bits that are genuinely random or pseudo-random, 
so that they can be used for probability sampling.
```

File: specification/trace/tracestate-probability-sampling.md [hunk 8]
```diff
@@ -295,432 +286,355 @@ applies to the two samplers specified here as well as consumers of
...
+- No implementations set the Random flag
+- All OpenTelemetry trace SDKs generate 128 random bits, except when there is a user-supplied IdGenerator
+- Some OpenTelemetry propagation SDK components do not propagate
+  unrecognized trace flags.
 
-Consumers SHOULD unset `p` from the `tracestate` if the unsigned
-decimal value is greater than 63 before using the `tracestate` to make
-a sampling decision or interpret adjusted count.
+These points, taken together, means there will be a long delay before
+the Random trace flag functions as intended in ther OpenTelemetry
```
Comment (kentquirk) [inline specification/trace/tracestate-probability-sampling.md:357]:
```suggestion
the Random trace flag functions as intended in the OpenTelemetry
```

File: specification/trace/tracestate-probability-sampling.md [hunk 9]
```diff
@@ -295,432 +286,355 @@ applies to the two samplers specified here as well as consumers of
...
+Head Samplers (i.e., trace SDKs) SHOULD by default generate the 56
+random bits specified for TraceIDs by the W3C Trace Context Level 2,
+and set the Random trace flag, when generating new contexts.
 
-#### Examples: Context invariants
+##### Requirement: Tail Samplers assume random TraceIDs
 
-##### Example: Probability sampled context
+Tail Samplers SHOULD ignore the W3C Trace Context Level 2 Random flag
+when interpreting span data.  This is a compromise
```
Comment (kentquirk) [inline specification/trace/tracestate-probability-sampling.md:372]:
```suggestion
when interpreting span data.  This is a compromise.
```

File: specification/trace/tracestate-probability-sampling.md [hunk 10]
```diff
@@ -295,432 +286,355 @@ applies to the two samplers specified here as well as consumers of
...
 
-#### Examples: Context invariants
+##### Requirement: Tail Samplers assume random TraceIDs
 
-##### Example: Probability sampled context
+Tail Samplers SHOULD ignore the W3C Trace Context Level 2 Random flag
+when interpreting span data.  This is a compromise 
 
-Consider a trace context with the following headers:
+When there is an explicit R-value set, the Sampler will anyway
```
Comment (kentquirk) [inline specification/trace/tracestate-probability-sampling.md:374]:
```suggestion
When there is an explicit R-value set, the Sampler will
```

**Review by kentquirk — COMMENTED** [review]
I'm a fan of this. I think you've done a good job here and based on what I see, I don't think we need to be rewriting it from scratch. Our past history leads me to believe that might be a long process. Let's push forward!

File: specification/trace/sdk.md [hunk 11]
```diff
@@ -312,21 +333,46 @@ The default sampler is `ParentBased(root=AlwaysOn)`.
...
+
+* Ratio values are restricted to the range `2**-56` through 1
+* A rejection threshold is calculated, expressing as an integer how
+  many out of 2**56 trace IDs should be sampled.
+* The threshold is encoded as a "T-value", expressing the threshold
+  using up to 5 hexadecimal digits of precision.
+* Sampler decisions are made by comparing the Trace ID randomness
+  against the rejection threshold.
+* When Sampled, T-value is included in the [OpenTelemetry TraceState
+  header](./tracestate-handling.md), identified by sub-key `th`,
```
Comment (kalyanaj) [inline specification/trace/sdk.md:359]: Should we update the tracestate-handling.md to include the examples for the sub-key 'th'?  Right now, it seems to be referring to the earlier p values and r value examples.

File: specification/trace/sdk.md [hunk 12]
```diff
@@ -312,21 +333,46 @@ The default sampler is `ParentBased(root=AlwaysOn)`.
...
+The implementation has the following steps:
+
+* Ratio values are restricted to the range `2**-56` through 1
+* A rejection threshold is calculated, expressing as an integer how
+  many out of 2**56 trace IDs should be sampled.
+* The threshold is encoded as a "T-value", expressing the threshold
+  using up to 5 hexadecimal digits of precision.
+* Sampler decisions are made by comparing the Trace ID randomness
+  against the rejection threshold.
+* When Sampled, T-value is included in the [OpenTelemetry TraceState
```
Comment (kalyanaj) [inline specification/trace/sdk.md:358]: It will be good rephrase this to an active voice of what implementations must do. For example, something like:  "When sampled, an implementation MUST update the tracestate header to specify the T-value. <+ any additional details>"
